### PR TITLE
[UR][L0] Implement bindless image functions from ZE_experimental_bindless_image

### PR DESCRIPTION
**Pull request by wenju-he**
Function zeMemGetPitchFor2dImage and zeImageGetDeviceOffsetExp from
L0 extension ZE_experimental_bindless_image are used.

In urBindlessImagesUnsampledImageCreateExp, if image description is
different from that of image allocated by urBindlessImagesImageAllocateExp,
an image view is created from the image.

### CODE REVIEW

**Comment by wenju-he** [comment]
@HoppeMateusz could you please review? thank you

**Comment by wenju-he** [comment]
@jsji @YuriPlyakhin could you please review, thank you.

File: source/adapters/level_zero/device.cpp
Comment (HoppeMateusz) [inline source/adapters/level_zero/device.cpp:926]: ZE experimental bindless image is only defined for 2D image - this is what cuda support - and we didnt decide to support 1D.

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [inline source/adapters/level_zero/image.cpp:609]: we do not support sampled images - shoudl this function return SUCCESS ?

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [inline source/adapters/level_zero/image.cpp:711]: we need BINDLESS FLAG here too - image from pitched memory must be created with bindless descriptor:

https://spec.oneapi.io/level-zero/latest/core/api.html#ze-image-pitched-exp-desc-t

struct ze_image_pitched_exp_desc_t
**Image descriptor for bindless images** created from pitched allocations

Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:711]: So it would be a chain of pNext like your test code:
```
    ze_image_pitched_exp_desc_t pitchedDesc = {};
    pitchedDesc.stype = ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC;
    pitchedDesc.ptr = reinterpret_cast<void *>(0x1234000);

    ze_image_bindless_exp_desc_t bindlessExtDesc = {};
    bindlessExtDesc.stype = ZE_STRUCTURE_TYPE_BINDLESS_IMAGE_EXP_DESC;
    bindlessExtDesc.pNext = &pitchedDesc;
    bindlessExtDesc.flags = ZE_IMAGE_BINDLESS_EXP_FLAG_BINDLESS;
```
ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC is only used for bindless image, right?
Alternatively can we set bindless flag to true when ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC  type is present? This would save us a pNext.
IMHO the chain of two pNext make more sense if ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC has multiple purposes.

`Image descriptor for bindless images created from pitched allocations` also indicates that ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC is only used for bindless image.

File: source/adapters/level_zero/image.cpp
Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:600]: @ph0b, bindless sampler isn't supported in IGC and L0 runtime. For instance, if there is an array of sampled images used in a kernel, the case can't work correctly. It is a technical blocker at the moment. There is an on-going discussion of the issue involving @HoppeMateusz and IGC people.

@HoppeMateusz the reason I return UR_RESULT_SUCCESS here is that I'm able to make SYCL bindless sampled image tests that only use sampled image as kernel argument, e.g [sampling_1D.cpp](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/bindless_images/sampling_1D.cpp) [sampling_2D.cpp](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/bindless_images/sampling_2D.cpp) and [sampling_3D.cpp](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/bindless_images/sampling_3D.cpp), working correctly on DG2, with the help of a few draft changes in IGC and L0 runtime. Basically sampler still needs a sampler_index encoded by IGC and L0 runtime saves sampler offset to a place according to sampler_index.

Return UR_RESULT_SUCCESS here also would allow us to test bindless sampler support as the above-mentioned technical blocker is in the process of being resolved.

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [reply source/adapters/level_zero/image.cpp:702]: yes, this has to be a chain.

For now, ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC  is only defined for BINDLESS images - so we need to pass  ze_image_bindless_exp_desc_t   too. 

There could be possible etnesion that will allow creating non-bindless images with pitched pointer - something like OpenCL's:

cl_mem clCreateImageWithProperties(
    cl_context context,
    const cl_mem_properties* properties,
    cl_mem_flags flags,
    const cl_image_format* image_format,
    const cl_image_desc* image_desc,
    void* host_ptr,
    cl_int* errcode_ret);

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [reply source/adapters/level_zero/image.cpp:600]: ok, thanks for explanation.

Wouldnt it be better to return error code from the function call urBindlessImagesUnsampledImageHandleDestroyExp() then ? Or Is it failing?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:633]: Should we try to share the code with `urMemImageCreate` in memory.cpp?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:622]: Did not see this in L0 spec example, is this required for bindless image only?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:652]: Similarly, should we try to share the code with `urMemRelease`?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:741]: Should we init it?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:600]: Do we need to call `zeSamplerDestroy` for `ZeSampler`?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:766]: Do we want to assert `hSampler->ZeSampler`  is not null?

File: source/adapters/level_zero/image.cpp
Comment (jsji) [inline source/adapters/level_zero/image.cpp:778]: Should we try to share some code with `enqueueMemImageCommandHelper` in memory.cpp?

File: source/adapters/level_zero/device.cpp
Comment (HoppeMateusz) [reply source/adapters/level_zero/device.cpp:926]: thanks

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [reply source/adapters/level_zero/image.cpp:621]: yes, memory resdiency of bindless images must be controlled by user of L0 driver , calling it here is correct.

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [inline source/adapters/level_zero/image.cpp:676]: is this correct ? 

ptr is expected to be raw pointer returned from  urUSMPitchedAllocExp -> urUSMDeviceAlloc(hContext, hDevice, pUSMDesc, pool, Size, ppMem) ->  ppMem

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [inline source/adapters/level_zero/image.cpp:728]: there is no mechanism to prevent wrong casts ?
 
we rely on application calling  XXXSampledImageXXX() function with only Sampled Images

and UnsampledImage - with non sampled images? 

pasign Unsampled image handle to this function will crash  / or be UB

File: source/adapters/level_zero/image.cpp
Comment (HoppeMateusz) [reply source/adapters/level_zero/image.cpp:743]: why cant we use Image function   to copy from HOST  to Bindless Image ?

ze_image_handle_t  returned from zeImageCreate should work with standard image API

Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:758]: if the underlying memory is non-usm, i.e. created by zeImageCreate, zeCommandListAppendImageCopyFromMemory is called as shown below, so it is image function.
I was intending to use zeCommandListAppendImageCopyFromMemoryExt/zeCommandListAppendImageCopyToMemoryExt, but then find they are not implemented in level-zero runtime. Is this a gap? @HoppeMateusz

File: source/adapters/level_zero/image.cpp
Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:743]: ur_exp_image_handle_t size is 64bit for unsampled image and 128bit for sampled image for L0 adaptor.
Unfortunately size info isn't passed to this function.
In the long run the ideal solution is likely to support 64bit sampled image handle. For instance, can we always set bindless sampler offset to be the same as bindless image offset when zeImageCreate accepts a sampler argument in the future like cuda api cuTexObjectCreate? Then only one 64bit handle is enough for sampled image.

File: source/adapters/level_zero/image.cpp
Comment (wenju-he) [reply source/adapters/level_zero/image.cpp:689]: hImageMem can be either an image created by zeImageCreate or usm created by zeMemAllocDevice. So a query is needed at the beginning of this function.
added a check and return error if MemAllocProperties.type isn't device usm

Comment (HoppeMateusz) [reply source/adapters/level_zero/image.cpp:691]: ok, 
it wasnt clear to me that  ur_exp_image_mem_handle_t   is interpreted as image handle or just a raw pointer.

**Review by HoppeMateusz — APPROVED** [review]

**Review by ph0b — APPROVED** [review]

File: source/adapters/level_zero/image.cpp
Comment (jsji) [reply source/adapters/level_zero/image.cpp:631]: Thanks @wenju-he  and @HoppeMateusz !

File: source/adapters/level_zero/image.cpp
Comment (jsji) [reply source/adapters/level_zero/image.cpp:748]: Sounds good.

**Review by jsji — APPROVED** [review]
Thanks!

**Review by nrspruit — APPROVED** [review]
LGTM

**Review by raiyanla — APPROVED** [review]

**Comment by kbenzie** [comment]
@nrspruit this should not have been merged without the approval of @oneapi-src/unified-runtime-bindless-images-write. It should also not have been merged before other PR's in the [ready to merge](https://github.com/oneapi-src/unified-runtime/pulls?q=is%3Aopen+is%3Apr+label%3A%22ready+to+merge%22) queue. Please reach out before pressing merge next time.

**Comment by kbenzie** [comment]
@nrspruit and there's always room to accomodate deliverable priorities while keeping in the queue.

**Pull request merged**
